src/kernel/thread.c

```diff
@@ -299,6 +300,22 @@ void doNBRecvFailedTransfer(tcb_t *thread)
     setRegister(thread, badgeRegister, 0);
 }
 
+static void prepareSetDomain(tcb_t *tptr, dom_t dom)
+{
+    if (config_set(CONFIG_HAVE_FPU)) {
+        /* Save FPU state now to avoid touching cross-domain state later */
+        fpuRelease(tptr);
+    }
```

Comment:
I'll keep on trying to find an alternative way of handling this but for now we do need it for verification, so thank you for adding it. However, I think it would be better with a check to avoid saving the fpu state if for whatever reason we're setting the domain of the thread to the current domain. That would help us avoid having to think through and reason about whether there's any code paths that lead to resuming a thread with its fpu state unexpectedly cleared.
```suggestion
    if (config_set(CONFIG_HAVE_FPU)) {
        if (ksCurDomain != dom) {
            /* Save FPU state now to avoid touching cross-domain state later */
            fpuRelease(tptr);
        }
    }
```

Reply:
Good find! Seems that that case is currently handled, but perhaps not intentionally:

setdomain calls rescheduleRequired, which set `ksSchedulerAction = SchedulerAction_ChooseNewThread`.
After handling a syscall, the call chain is schedule -> scheduleChooseNewThread ->  chooseThread which calls either switchToThread or switchToIdleThread.

However, the other call to switchToThread in schedule has `assert(candidate != NODE_STATE(ksCurThread))`, which to me hints it's an unexpected case. Then again, the assert isn't in switchToThread itself.

Anyway, wouldn't it be better to do the `ksCurDomain != dom` check in decodeDomainInvocation and skip calling setdomain altogether?

Reply:
Or an `if (tptr->tcbDomain == dom) { return; }` in setDomain itself.

Reply:
Oh, good thinking! The second approach of doing the check and returning in `setDomain` would be better for verification, or at least require less other changes. There are some technicalities with our specifications that I'm a bit concerned about though, so it might be best to leave this for after the current deadline for when I have time to explore it in more detail.